### ISSUE TITLE
Custom Shop

### ISSUE THREAD

**Issue by l34Um1** [OpenAngelArena/oaa#2093]
# Custom Shop

A custom shop with 6 pages. Should be written to support more or less pages as well.

The first page is “Basics” and includes everything that is currently listed in the “Basics” shop tap.
The second page is “Tier 1” and includes all tier 1 versions of items. E.g. Aghanims, Heart, Skadi, Force Staff etc.
The third page is “Tier 2” and includes all tier 2 items. E.g. Aghanims 2, Heart 2, Skadi 2, Force Staff 2, Pull Staff 1, etc.
The fourth page is “Tier 3” and includes all tier 3 items. E.g. Aghanims 3, Heart 3, Regen Crystal 1, Skadi 3, Force Staff 3, Pull Staff 2, etc.
The fifth page is “Tier 4” and includes all tier 4 items. E.g. Aghanims 4, Heart 4, Regen Crystal 2, Heart Transplant 1, Skadi 4, Covfefe 1, Force Staff 4, Pull Staff 3, etc.
The sixth page is “Tier 5” and includes all tier 5 items. E.g. Aghanims 5, Heart 5, Regen Crystal 3, Heart Transplant 2, Skadi 5, Covfefe 2, Force Staff 5, Pull Staff 4, etc.

The items should be visually grouped in such a way that branches are placed close to their origin.

The shop needs to keep all current functionality, like pinning, queuing, “Show in Shop”, search, etc.
The shop also needs to be easily customisable for us in terms of item location.

### Inspect Upgrade Tree
Further the shop needs a new feature: “Inspect Upgrade Tree”. This function is placed in the bottom area of the shop. When an item is selected the "Inspect Upgrade Tree" button appears. When rightclicking an item in your inventory you can also select this option.

![unbenannt-2](https://user-images.githubusercontent.com/24982519/37799422-ed907f3e-2e1f-11e8-8b81-09ddc02bc6f1.jpg)

When selected a new window expands out of the left side of the shop, similar to item guides. In this window all possible upgrade paths for that particular item are shown, connected via lines. 
 
Heart as an example:
![](https://lh5.googleusercontent.com/zAP2AB1Z5XkkY02Y8TcY84wq-KEKaf0_Jmbq0HElbq7GporgOfi0i3ueiZ3B9t7Y1Mly9tUkSTj4_vIJXVjMO0r_ePPN_XfiqgqJ3hTblv3q15UQm_WVz-a5C51hDIjH-xzRV4W2)

It should show icons instead of just text. Selecting any of the items in here will show the item as selected in the regular shop interface. Items can also be purchased here.

[Bug Bounty Guideline](https://docs.google.com/document/d/1Tpm0j1ekVc6X-LRlF_6Q81RMGVlOOO5EBvSf_xAQk4I/edit)

**Comment by chrisinajar** [comment]
A few things about the implementation of this shop, you should expose an API for interacting with the shop so that other features can be added on top of this construct. Additionally, items should be able to be purchased with non-gold resources (easy with custom shop, doable with purchase filters for a modified shop) and some items should appear in the shop but not be purchasable.

If you modify the existing shop in-place you'll need to expose an API for loading state into it, if you create the shop from scratch (preferred) then such an API should intrinsically be part of your implementation.

This is a *very hard ticket* because matching existing dota functionality is very challenging. There are lots of custom shops out there but all of them act weird in one way or another. If we load in the normal items and disable the new side panel, end users should not be able to tell the shop is custom.

We have 3 shops currently and plan on adding a 4th.

**Comment by ZLOY5** [comment]
Hi!
I think I can done this, but first I want figure out what you expect to see in the end. You need fully new system with custom shop triggers, shop-entity and etc. or  just redesigned UI for current dota shops?

Can you give examples of needed API?

**Comment by ZLOY5** [comment]
Sorry me for long absense.
You can see progress in my fork or this low-resolution screenshot.

And can anyone give me item list by tiers?

![shot_0357](https://user-images.githubusercontent.com/10956575/38826159-b06cdaa2-41b7-11e8-8bae-ae622f73ea8d.jpg)

**Comment by l34Um1** [comment]
\oaa\game\scripts\npc\items

is probably the best you'll find. Every item is in there with a KV file for each tier.

**Comment by fcalife** [comment]
I'd like to know if this is still claimed or has opened up. I'd like to at least give it a shot if so.

**Comment by l34Um1** [comment]
So I tried to test this but I'm just getting parsing errors and when I'm ingame there is no shop at all. Can you post some screenshots or even better a video of the progress you have made?

**Comment by ZLOY5** [comment]
Can you give me a log with errors?
I`ll try to make video

**Comment by ZLOY5** [comment]
https://streamable.com/ey8kh

**Comment by l34Um1** [comment]
Looks good. Although can you make the switching between pages not a dropdown menu but instead multiple buttons at the top?

**Comment by carlosrpg** [comment]
Adding a few user test scenarios:
 - Show in Shop should open shop on the correct tab.
 - consider that you have money to buy all ingredients for a lvl 2 and lvl 3 item, and you have a tier 3 and tier 4 cores. you should be able to buy it by right click the item on the shop directly.

**Comment by fcalife** [comment]
Just want to say congratulations, looks on the way to a job well done!

**Comment by DoctorGester** [comment]
Looks very cool, I'm glad someone took on it. Just be sure to fix your code style (indentation etc). Also imo this code is asking for unit tests because it is a huge codebase.

**Comment by l34Um1** [comment]
Here are the 3 errors I get when loading up your branch. I don't see any shop when loaded in.
![error 1](https://user-images.githubusercontent.com/24982519/39859923-c33564ce-543b-11e8-8f73-83c537c4c4fd.PNG)
![error 2](https://user-images.githubusercontent.com/24982519/39859924-c355a036-543b-11e8-834b-9a26549ae873.PNG)
![error 3](https://user-images.githubusercontent.com/24982519/39859926-c3704d32-543b-11e8-9712-846da474cd0b.PNG)

**Comment by DoctorGester** [comment]
@l34Um1 interestingly enough first two errors are in dota's own styles and in fact my friend has at least the error on the first screenshot in his own project.

**Comment by ZLOY5** [comment]
@l34Um1 fixed I think
P.S.: recompile oaa_test.vmap if you want buy items not only to stash

**Comment by l34Um1** [comment]
Yeah I can see the shop now. Good work so far but still a bit to go.

**Comment by darklordabc** [comment]
Looking good!

**Comment by mikibun** [comment]
Is this still being worked on? @ZLOY5

**Comment by zhaog100** [comment]
Claiming this bounty. I'll implement the Custom Shop feature.

**Comment by DarkoniusXNG** [comment]
> Claiming this bounty. I'll implement the Custom Shop feature.

I do not know who will pay you even if you do it. Because Baumi left (and stopped playing dota 2) so he will not pay you.
